src/simulators/qasm/qasm_controller.hpp

```diff
@@ -379,31 +402,53 @@ OutputData QasmController::run_circuit(const Circuit &circ,
 // Utility methods
 //-------------------------------------------------------------------------
 
-QasmController::Method QasmController::simulation_method(const Circuit &circ) const {
-  // Check conditions for automatic simulation types
-  auto method = simulation_method_;
-  if (method == Method::automatic) {
-    // Check if Clifford circuit and noise model
-    if (validate_state(Stabilizer::State(), circ, noise_model_, false)) {
-      method = Method::stabilizer;
-    } else {
-    // Default method is statevector, unless the memory requirements are too large
+QasmController::Method
+QasmController::simulation_method(const Circuit &circ,
+                                  const Noise::NoiseModel &noise_model,
```

Comment:
One of the problems that I see is that the `noise_model` is now traveling around across functions that should know nothing about noise models, for example this member function, needs the noise_model just because there's a validation that requires it. But the only purpose of this member function is to return the simulation method, there's no need for the noise model at all.
I'd move out validation from this member function (so the noise_mode argument). Haven't seen where to move it actually but I saw that we are already calling validation just before running a circuit, so this might be enough?

Reply:
Actually the simulation method *does* critically depend on both the noise model and the circuit. You can only use the stabilizer method if both the circuit and noise model is Clifford for example. This was being checked before, but was using the noise model stored as a controller member variable

Reply:
Ok then.

Reply:
Ok then!